### PR TITLE
Fixes #28554 -- Added support for multiple files for forms.FileField

### PR DESCRIPTION
**Pull request by codingjoe**
https://code.djangoproject.com/ticket/28554

### CODE REVIEW

**Comment by FlipperPA** [comment]
I love the idea for the feature, but would echo the comment from the Google Groups thread: rather than passing `multiple` as an argument, would a `forms.MultiFileField` be cleaner and avoid a lot of `if / else` code blocks?

**Comment by codingjoe** [comment]
@FlipperPA I understood the comment to be addressing `models.MultiFielField` not the `forms` implementation. But maybe I got this wrong.

Anyhow, let me explain my reasoning. Maybe it makes sense, maybe it doesn't.
We are talking about forms here, and this is very close to how html forms work. There is no
```html
<input type="multifile">
```
but a
```html
<input type="file" multiple>
```
That is why this seemed a more obvious solution to me. Furthermore there is already documentation on [how to upload multiple files in Django](https://docs.djangoproject.com/en/1.11/topics/http/file-uploads/#uploading-multiple-files). The multiple attribute was favored too. It would seem somewhat inconsistent.

I would even go as far as saying that the `FileInput` should always return a list of files. I just fear that this would be too much of an API change. I will take this discussion back to the mailing list.

In any case, thank you so much for the involvement and feedback. I really appreciate it and will try to do good by everyone :)

thx
-joe

File: django/forms/fields.py [hunk 1]
```diff
@@ -532,10 +532,13 @@ class FileField(Field):
...
     }
 
-    def __init__(self, *, max_length=None, allow_empty_file=False, **kwargs):
+    def __init__(self, *, max_length=None, allow_empty_file=False, multiple=False, **kwargs):
         self.max_length = max_length
         self.allow_empty_file = allow_empty_file
+        widget = kwargs.get('widget', self.widget)
         super().__init__(**kwargs)
+        if isinstance(widget, type):
+            self.widget.multiple = multiple
```
Comment (melvyn-sopacua) [inline django/forms/fields.py]: Confusing bit. If superclass' __init__ changes self.widget then what?

File: django/forms/widgets.py [hunk 2]
```diff
@@ -373,13 +374,43 @@ class FileInput(Input):
...
+                multiple = False
+            else:
+                warnings.warn(
+                    'The attr "multiple" is deprecated in favor of the keyword argument "multiple".',
+                    RemovedInDjango30Warning,
+                )
+        self.multiple = multiple
+        if self.multiple:
+            attrs.setdefault('multiple', True)
+        super().__init__(attrs=attrs)
```
Comment (melvyn-sopacua) [inline django/forms/widgets.py]: if I pass `attrs={'multiple': False)` and init with multiple=True, they are out of sync.

File: django/forms/fields.py (hunk 1 above)
Comment (codingjoe) [reply django/forms/fields.py]: @melvyn-sopacua yes, it really is.
Super actually does change `self.widget` but it doesn't have a concept of `multiple` nor of any other argument passed to the widget.

I think a comment could go a long way here, what do you think?

File: django/forms/widgets.py (hunk 2 above)
Comment (codingjoe) [reply django/forms/widgets.py]: @melvyn-sopacua that assessment is correct. If you'd to write
```python
ClearableFileInput(multiple=True, attrs={'multiple': False})
```
you would end up with a HTML input that only allows one file, but you would get a list containing 0..1 file in Django. I agree the use case for that is slim. I could raise a warning, what do you think?

**Comment by codingjoe** [comment]
@melvyn-sopacua thanks for taking your time to review this 👍 I really appreciate the help and effort.

**Comment by codingjoe** [comment]
@melvyn-sopacua I took your feedback and reviewed how other attributes such as `widget.is_required` are set. It is more strict and overwrite any behavior specified by the widget.
I adopted that simplicity and reduced the complexity a bit. The behavior should be more straight forward now. Again thanks for the feedback.

**Comment by codingjoe** [comment]
@timgraham I know it's short notice, but maybe we can get this still in 2.0.

File: django/forms/fields.py [hunk 3]
```diff
@@ -575,6 +577,8 @@ def clean(self, data, initial=None):
             data = None
         if not data and initial:
             return initial
+        if self.multiple:
+            return [super(FileField, self).clean(value) for value in data]
```
Comment (grunskis) [inline django/forms/fields.py]: Why not just `super().clean(value)`?

File: django/forms/widgets.py [hunk 4]
```diff
@@ -410,27 +436,36 @@ def is_initial(self, value):
...
     def get_context(self, name, value, attrs):
         context = super().get_context(name, value, attrs)
         checkbox_name = self.clear_checkbox_name(name)
         checkbox_id = self.clear_checkbox_id(checkbox_name)
+        is_initial = self.is_initial(value)
         context['widget'].update({
             'checkbox_name': checkbox_name,
             'checkbox_id': checkbox_id,
-            'is_initial': self.is_initial(value),
+            'is_initial': is_initial,
```
Comment (grunskis) [inline django/forms/widgets.py]: Was this intentional? Doesn't seem to be related to the task...

File: docs/releases/2.0.txt [hunk 5]
```diff
@@ -207,6 +207,12 @@ Forms
   <django.forms.Form.errors.get_json_data>` method returns form errors as
   a dictionary suitable for including in a JSON response.
 
+* :class:`~django.forms.FileField` as well as :class:`~django.forms.FileInput`
+  and :class:`~django.forms.ClearableFileInput` support a new ``multiple``
+  argument. It allows uploading multiple files using a single file input.
+  In favor of the new argument ``attrs={'multiple': True}`` will emit a
+  warning.
```
Comment (grunskis) [inline docs/releases/2.0.txt]: Last sentence isn't very clear. It contains 2 different ideas and could be made more clear by splitting it into 2 sentences.

File: tests/forms_tests/field_tests/test_filefield.py [hunk 6]
```diff
@@ -80,3 +80,32 @@ def test_disabled_has_changed(self):
...
+
+    def test_multiple(self):
+        f = FileField(multiple=True)
+        files = [SimpleUploadedFile('file_%s' % i, ('%s' % i).encode('utf-8')) for i in range(2)]
+        self.assertEqual(files, f.clean(files))
+
+        # ensure validation is executed for all files
+        f = FileField(multiple=True)
+        files = [SimpleUploadedFile('file_%s' % i, b'') for i in range(2)]
+        with self.assertRaisesMessage(ValidationError, 'The submitted file is empty.'):
```
Comment (grunskis) [inline tests/forms_tests/field_tests/test_filefield.py]: Would be handy if this tells which was was empty. Or change the message to - "One of the submitted files was empty".

File: tests/forms_tests/field_tests/test_filefield.py [hunk 7]
```diff
@@ -80,3 +80,32 @@ def test_disabled_has_changed(self):
...
+            f.clean(files)
+
+        def test_validator(value):
+            if not isinstance(value, SimpleUploadedFile):
+                raise ValidationError('Value is not a file.')
+
+        f = FileField(multiple=True, validators=[test_validator])
+        files = [SimpleUploadedFile('file_%s' % i, ('%s' % i).encode('utf-8')) for i in range(2)]
```
Comment (grunskis) [inline tests/forms_tests/field_tests/test_filefield.py]: Didn't you want to actually test that the validator raises `ValidationError` here?

File: tests/forms_tests/widget_tests/test_clearablefileinput.py [hunk 8]
```diff
@@ -31,6 +32,22 @@ def test_clear_input_renders(self):
...
+            """
+            Currently: <a href="something">something</a>
+            <input type="checkbox" name="myfile-clear" id="myfile-clear_id" />
+            <label for="myfile-clear_id">Clear</label><br />
+            Change: <input multiple type="file" name="myfile" />
+            """
+        ))
+        self.check_html(self.widget_cls(multiple=True), 'myfile', [FakeFieldFile(), FakeFieldFile()], html=(
+            """
+            Currently: 2 files
```
Comment (grunskis) [inline tests/forms_tests/widget_tests/test_clearablefileinput.py]: Why not showing list of links here? Like this you don't actually know what files you're clearing...

File: django/forms/fields.py (hunk 3 above)
Comment (codingjoe) [reply django/forms/fields.py]: You can't call super inside a generator without a specific target and instance. I will make sure to add a comment.

File: tests/forms_tests/field_tests/test_filefield.py (hunk 6 above)
Comment (codingjoe) [reply tests/forms_tests/field_tests/test_filefield.py]: That is a good point. It will make the implementation more complex tho. What do you think? Worth it or not?

File: tests/forms_tests/field_tests/test_filefield.py (hunk 7 above)
Comment (codingjoe) [reply tests/forms_tests/field_tests/test_filefield.py]: oh, yes good catch

File: tests/forms_tests/widget_tests/test_clearablefileinput.py (hunk 8 above)
Comment (codingjoe) [reply tests/forms_tests/widget_tests/test_clearablefileinput.py]: That was my first thought too, but it is not how browsers do it for the input itself. UX wise this will always look consistent even if a user selects thousands of files. Murphy's Law, someone will do it ;)

**Comment by srkunze** [comment]
What's the status of this pull request? :-)

**Comment by codingjoe** [comment]
@srkunze a review would certainly help :) I am just waiting for feedback.

File: tests/forms_tests/field_tests/test_filefield.py [hunk 9]
```diff
@@ -80,3 +80,35 @@ def test_disabled_has_changed(self):
...
+
+        f = FileField(multiple=True, validators=[test_validator])
+
+        files = [SimpleUploadedFile('file_1', bytes(range(100))),
+                 SimpleUploadedFile('file_1', b'foo')]
+        with self.assertRaisesMessage(ValidationError, 'File too large.'):
+            f.clean(files)
+
+    def test_multiple_widget(self):
+        self.assertTrue(FileField(multiple=True).widget.multiple)
```
[the remark this answers is not in the document]
Comment (codingjoe) [reply tests/forms_tests/field_tests/test_filefield.py]: Is the `assertTrue` deprecated or am I missing something? I happy to make the change, but I also want to understand the logic behind it, to not repeat the mistake.

[the remark this answers is not in the document]
Comment (charettes) [reply tests/forms_tests/field_tests/test_filefield.py]: @codingjoe the issue with `assertTrue` is that it's more of a `assertTruthy`; any value that evaluates to `True` will pass (e.g. `1`, `'hello'`, ...).

[the remark this answers is not in the document]
Comment (timgraham) [reply tests/forms_tests/field_tests/test_filefield.py]: `assertTrue(value)` will pass for `bool(value) is True` which is different than checking for `True`.

**Comment by codingjoe** [comment]
@timgraham I dropped the warning and made sure it's backwards compatible.
Having `attrs={'multiple'=True}` will not enable this feature and not change the return value of `value_from_datadict`. Only setting `multiple=True` as a keyword argument enables the new feature.

I would add a new warning tho to inform people that `attrs={'multiple'=True}` isn't the same as doing `multiple=True`.

**Comment by timgraham** [comment]
The documentation should be more descriptive about how `multiple=True` changes the field's functionality and how it differs from `attrs={'multiple': True}`. However, I'm not sure this is a good design though (offering two seeming similar ways to achieve some functionality that work subtly differently). Probably this discussion should continue on the mailing list.

**Comment by codingjoe** [comment]
@timgraham ok, than it certainly wont make it into this release. I will take it with me to DjangoConEU and discuss it with a couple people there.

File: django/forms/widgets.py [hunk 10]
```diff
@@ -364,13 +364,30 @@ class FileInput(Input):
     needs_multipart_form = True
     template_name = 'django/forms/widgets/file.html'
 
+    def __init__(self, attrs=None, multiple=False):
+        if attrs is None:
+            attrs = {}
+        self.multiple = multiple
+        if multiple:
+            attrs.setdefault('multiple', multiple)
```
[the remark this answers is not in the document]
Comment (codingjoe) [reply django/forms/widgets.py]: Good point. I could just force it to be true then.

**Comment by codingjoe** [comment]
@adamchainz there are changes made to `docs/topics/http/file-uploads.txt` ;)

File: tests/forms_tests/field_tests/test_filefield.py (hunk 6 above)
Comment (codingjoe) [reply tests/forms_tests/field_tests/test_filefield.py]: Validation is currently done per file base. That is something that I don't want to change. I will add the file name to the message to be more verbose. Caveat: There could be multiple empty files. Since we raise `ValidationErrors` only the first one will be shown. That is the same for all Validators tho. I don't think this is really unexpected.

File: docs/topics/http/file-uploads.txt [hunk 11]
```diff
@@ -127,6 +127,8 @@ field in the model::
             form = UploadFileForm()
         return render(request, 'upload.html', {'form': form})
 
+.. _topic-multi-file-upload:
+
 Uploading multiple files
 ------------------------
```
Comment (adamchainz) [inline docs/topics/http/file-uploads.txt]: line 135

```
If you want to upload multiple files using one form field, set the ``multiple`` HTML attribute of field's widget:	 HTML attribute of field's widget:
```

->

```
If you want to upload multiple files using one form filed, pass ``multiple=True`` to :class:`FileField`:
```

(but get the `:class:` ref right)

**Comment by codingjoe** [comment]
Ok, this works now. There is nothing left todo unless someone has some final concerns.

**Comment by codingjoe** [comment]
@timgraham any chance to get this in before the feature freeze?

File: tests/forms_tests/field_tests/test_filefield.py [hunk 12]
```diff
@@ -80,3 +80,34 @@ def test_disabled_has_changed(self):
...
+            if value.size > 10:
+                raise ValidationError('File too large.')
+
+        f = FileField(multiple=True, validators=[test_validator])
+
+        files = [
+            SimpleUploadedFile('file_1', bytes(range(100))),
+            SimpleUploadedFile('file_1', b'foo')
+        ]
+        with self.assertRaisesMessage(ValidationError, 'File too large.'):
```
Comment (timgraham) [inline tests/forms_tests/field_tests/test_filefield.py]: Is UX lacking here? Is there an indication of which file is too large?

**Comment by collinanderson** [comment]
I also personally think a separate class would be good (similar to SelectMultiple). It has the added advantage of being able to easily customize the templates for just just the multi files classes. I'm imagining we could even keep the `.multiple` attribute under the hood, so it could be as simple as something like:

```
class MultiFileField(FileField):
    multiple = True
```

So the actual code wouldn't need to change too much.

I agree that being able to remove individual files in ClearableMultipleFileInput isn't totally necessary right away. (though could be frustrating to users)

File: django/forms/fields.py [hunk 13]
```diff
@@ -531,18 +531,20 @@ class FileField(Field):
     default_error_messages = {
         'invalid': _("No file was submitted. Check the encoding type on the form."),
         'missing': _("No file was submitted."),
-        'empty': _("The submitted file is empty."),
+        'empty': _("The submitted file '%(file_name)s' is empty."),
```
[the remark this answers is not in the document]
Comment (codingjoe) [reply django/forms/fields.py]: I just tried the approach you suggested. It highly increases complexity. The `to_python` method, that raises the validation errors is only handed a single file at a time. Thus it doesn't know the difference between one or multiple files. If you don't have any strong feelings about this, I would prefer to keep it as is.

File: tests/forms_tests/field_tests/test_filefield.py (hunk 12 above)
Comment (codingjoe) [reply tests/forms_tests/field_tests/test_filefield.py]: This is a custom validation, but I modified the test to be more precise.

**Comment by codingjoe** [comment]
> Thanks for your effort here. I see you've been working on this for a long time.
Still, I have to say I'm a bit 😬 at this point.

Yes. It took awhile to come to a consensus on the mailing list and during sprints.

> Getting multiple return types from FileInput.value_from_datadict() is
surprising, shall we say.

How so, that is how the file input works in HTML too. The concept in Django that file fields can only store a single file, is somewhat wrong if you ask me. The whole point of this ticket was to be only the first step of actually adding multi file support to the ORM. It appeared wiser at the time, to add support to the form layer first, before implementing it into the ORM.

> Given that, the existing approach, using attrs and using files.getlist()
directly doesn't seem so bad. (The new code would not be significantly simpler.)

Currently you need to access the request layer, instead of the form layer. This approach is less complex especially if use the `ArrayField`. I didn't use the example in the documentation, because it is a PostgreSQL only feature. In this case you use a regular modelform to implement multi file support.

> A little-bit of weirdness may be OK, if there were clear gains elsewhere, say in
the rendering, but that doesn't seem to be the case. E.g. ClearableFileInput
gives only the count of files, whereas I'd hoped for a list of the current files;
having files individually clearable would also seem worthwhile… (I know you
asked on the mailing list about that and didn't get much input) and so on.

Yes, I did discuss that on the mailing list. You can not display the file names for multiple files, since it could be 1000 or more files. This is why all browser inputs type file only show the count for multiple files as well. The UX is consistent with browsers.

> Maybe we could accept the functionality as-is and then work on it over
time--that's got to be fair otherwise we never get anything in...--but I'm not
sure we're not better-off with a separate field class. (Despite there being no
<input type="multifile">: I don't think much follows from that.)

I also don't see a benefit of a separate class. There is only one file input in HTML, there is however a multiple attribute. Not only for `input=file` in fact, but also others.

> The third-party package linked from the ticket looks like the right kind of API.
(It doesn't offer the clearable widgets AFAICS, so wouldn't solve those
concerns.)

IMHO it does too many things at once and doesn't fulfil my purpose adding `multiple` support as defined in HTML.

> Given that my view logic has to change, selecting a different form field
doesn't seem to me to be too big an ask. Ultimately I don't see the benefit of
squeezing the dual functionality into the single class.

Ok, I'll write a longer post about this...

**Comment by codingjoe** [comment]
@carltongibson @collinanderson thanks for your feedback, let me try to explain a bit more the reasoning why I choose a kwarg and not a separate class.

I want to make sure that Django Forms and it's inputs follow the logic layed out in the HTML standard. New Django users should not need to learn a completely new concept. I believe you agree with this in all places where this is feasible.

Let's take a look into HTML then: HTML defines a set of different types. All of these types are usually represented by a single Django widget. The only place where Django drives here, is the clearablefileinput, since HTML has no concept of initial files. I find this still questionable.
There seems to be a one2one relation between HTML input type and Django input classes, that I don't want to break.

However, there are other HTML also defines the `multiple` attributen, see also: https://www.w3.org/TR/html52/sec-forms.html#element-attrdef-input-multiple

In fact `multiple` is defined for ALL input types. Django will need to address this sooner or later and I don't think we can add new class for every new attribute introduced. This would result in an exponential growth of the number of widget classes.

I wanted to avoid this rabbit hole, this is why I picked a kwarg. Anyhow, I can totally change the implementation, but I would rather deprecate and remove all multi value widgets and add a `multiple` argument to them too. Not as part of this commit of course and not in this version either.

**Comment by charettes** [comment]
> In fact multiple is defined for ALL input types. Django will need to address this sooner or later and I don't think we can add new class for every new attribute introduced. This would result in an exponential growth of the number of widget classes.

Even [if it is defined for all HTML input](https://html.spec.whatwg.org/multipage/input.html#the-multiple-attribute) it's only supported for handful of them (email, file) and not planed to be implemented for others by current browsers so that's a completely theoretical issue IMO. 

I'm in agreement with @carltongibson that we should adhere to the current `multiple` class breakdown to be coherent with the existing API. This is something we could revisit in the future through a deprecation period but as things stand right now I think using a `kwarg` for `FileInput` and different classes for `Select` only makes for an incoherent API.

**Comment by codingjoe** [comment]
So, I am hearing new class? No problem. I will however change the release notes to 3.0 I don't see this being merged before the freeze.

**Comment by codingjoe** [comment]
Just so we are on the same page, this would mean adding the following new classes:
* MultipleFileInput
* MultipleClearableFileInput
* MultipleFileField
* MultipleImageField
* MultipleAdminFileWidget

**Comment by codingjoe** [comment]
/ping @charettes just to confirm, that is what you are looking for correct? It's just a lot of work and I don't want to redo it yet again.

File: docs/releases/2.2.txt [hunk 14]
```diff
@@ -162,7 +162,12 @@ File Uploads
 Forms
 ~~~~~
 
-* ...
+* :class:`~django.forms.FileField` as well as :class:`~django.forms.FileInput`
```
Comment (auvipy) [inline docs/releases/2.2.txt]: move this to 3.0

File: docs/ref/forms/fields.txt [hunk 15]
```diff
@@ -606,10 +606,13 @@ For each field, we describe the default widget used if you don't specify
     * Error message keys: ``required``, ``invalid``, ``missing``, ``empty``,
       ``max_length``
 
-    Has two optional arguments for validation, ``max_length`` and
-    ``allow_empty_file``. If provided, these ensure that the file name is at
-    most the given length, and that validation will succeed even if the file
-    content is empty.
+    .. versionchanged:: 2.2
```
Comment (auvipy) [inline docs/ref/forms/fields.txt]: 3.0

File: docs/topics/http/file-uploads.txt [hunk 16]
```diff
@@ -127,21 +127,23 @@ field in the model::
...
     :caption: forms.py
 
     from django import forms
 
     class FileFieldForm(forms.Form):
-        file_field = forms.FileField(widget=forms.ClearableFileInput(attrs={'multiple': True}))
+        file_field = forms.FileField(widget=forms.ClearableFileInput(multiple=True))
 
-Then override the ``post`` method of your
+Then override the ``from_valid`` method of your
```
Comment (claudep) [inline docs/topics/http/file-uploads.txt]: Typo: `form_valid`

**Comment by AMK9978** [comment]
Due to inactivity, this PR could be closed. Given the changes in the [`FileField`](https://github.com/django/django/commit/fb4c55d9ec4bb812a7fb91fa20510d91645e411b), Not sure whether still supporting multiple files in one form field is needed or the original ticket needs to change to "someday".
